File: android/app/src/main/java/com/droppix/app/net/TransportClient.kt
package com.droppix.app.net

import com.droppix.app.protocol.MessageParser
import com.droppix.app.protocol.MsgType
import com.droppix.app.protocol.Protocol
import com.droppix.app.stats.RateMeter
import com.droppix.app.stats.StatsSink
import java.io.InputStream
import java.io.OutputStream
import java.net.InetSocketAddress
import java.security.cert.X509Certificate
import javax.net.ssl.SSLSocket

interface StreamListener {
    fun onConfig(config: Protocol.Config)
    fun onVideo(video: Protocol.Video)
    fun onAudio(pcm: ByteArray) {}
    fun onOverlay(show: Boolean) {}
}

class TransportClient {
    private val sendLock = Any()
    @Volatile private var out: java.io.OutputStream? = null
    // app->host sends originate on the UI thread (touch, orientation); socket I/O there
    // throws NetworkOnMainThreadException. Serialize every send onto one background thread.
    private val sender = java.util.concurrent.Executors.newSingleThreadExecutor { r ->
        Thread(r, "droppix-send").apply { isDaemon = true }
    }
    fun close() { sender.shutdown() }

    private fun longToBytes(x: Long) = ByteArray(8) { i -> (x ushr (56 - i * 8)).toByte() }
    private fun bytesToLong(b: ByteArray): Long {
        var x = 0L; for (i in 0 until 8) x = (x shl 8) or (b[i].toLong() and 0xFF); return x
    }

    // Thread-safe: called from the UI thread while run() reads on the net thread.
    fun sendTouch(contacts: List<com.droppix.app.protocol.Contact>) {
        val o = out ?: return
        val msg = Protocol.encodeMessage(MsgType.TOUCH, Protocol.encodeTouch(contacts))
        submitSend(o, msg)
    }

    fun sendOrientation(code: Int) {
        val o = out ?: return
        val msg = Protocol.encodeMessage(MsgType.ORIENTATION, Protocol.encodeOrientation(code))
        submitSend(o, msg)
    }

    // Encode on the caller; write on the background sender so UI-thread callers don't
    // hit NetworkOnMainThreadException. sendLock still serializes against net-thread writes.
    private fun submitSend(o: java.io.OutputStream, msg: ByteArray) {
        try {
            sender.execute {
                synchronized(sendLock) {
                    try { o.write(msg); o.flush() } catch (_: Exception) { /* dropped; loop will close */ }
                }
            }
        } catch (_: java.util.concurrent.RejectedExecutionException) { /* sender shut down */ }
    }

    // Wi-Fi/localhost path: open a TLS socket (pinning the host cert for non-localhost), then
    // run the protocol over its streams.
    fun run(host: String, port: Int, width: Int, height: Int, density: Int,
            fps: Int, audioWanted: Int, orientationCode: Int,
            listener: StreamListener, isRunning: () -> Boolean,
            stats: StatsSink? = null, pingIntervalMs: Long = 1000,
            name: String = "", id: String = "", tlsTrust: TlsTrust) {
        var serverCert: X509Certificate? = null
        val socket = tlsTrust.socketFactory { cert -> serverCert = cert }.createSocket() as SSLSocket
        try {
            socket.tcpNoDelay = true
            socket.connect(InetSocketAddress(host, port), 5000)
            socket.startHandshake()

            if (host != "127.0.0.1") {
                val cert = serverCert ?: throw IllegalStateException("no server certificate captured")
                val fp = certFingerprint(cert)
                val pinned = tlsTrust.pinnedFp(host)
                when {
                    pinned == null -> throw IllegalStateException("not paired")
                    pinned != fp -> throw CertChangedException(host)
                }
            }

            socket.soTimeout = 1000  // periodic wakeups so isRunning() is checked
            runOverChannel(socket.getInputStream(), socket.getOutputStream(),
                width, height, density, fps, audioWanted, orientationCode,
                listener, isRunning, stats, pingIntervalMs, name, id)
        } finally {
            try { socket.close() } catch (_: Exception) {}
        }
    }

    // Transport-agnostic protocol loop: sends HELLO, then reads CONFIG/VIDEO/AUDIO/OVERLAY and
    // answers PING while isRunning(). Used over a TLS socket (run) or a USB-accessory FD (AOA).
    // The caller owns `input`/`output` and closes them; this only nulls `out` on exit.
    fun runOverChannel(input: InputStream, output: OutputStream, width: Int, height: Int, density: Int,
                       fps: Int, audioWanted: Int, orientationCode: Int,
                       listener: StreamListener, isRunning: () -> Boolean,
                       stats: StatsSink? = null, pingIntervalMs: Long = 1000,
                       name: String = "", id: String = "") {
        try {
            out = output
            synchronized(sendLock) {
                output.write(Protocol.encodeMessage(MsgType.HELLO,
                    Protocol.encodeHello(Protocol.VERSION, width, height, density, name, id,
                        fps, audioWanted, orientationCode)))
                output.flush()
            }

            val parser = MessageParser()
            val chunk = ByteArray(65536)
            val frameRate = RateMeter(1000)
            var lastPing = 0L
            while (isRunning()) {
                val nowMs = System.currentTimeMillis()
                if (stats != null && nowMs - lastPing >= pingIntervalMs) {
                    synchronized(sendLock) {
                        output.write(Protocol.encodeMessage(MsgType.PING, longToBytes(System.nanoTime())))
                        output.flush()
                    }
                    lastPing = nowMs
                }
                val n = try { input.read(chunk) } catch (e: java.net.SocketTimeoutException) { 0 }
                if (n > 0) {
                    parser.feed(chunk, n)
                    var msg = parser.next()
                    while (msg != null) {
                        when (msg.type) {
                            MsgType.CONFIG -> Protocol.decodeConfig(msg.body)?.let(listener::onConfig)
                            MsgType.VIDEO -> {
                                Protocol.decodeVideo(msg.body)?.let(listener::onVideo)
                                if (stats != null) {
                                    frameRate.mark(System.currentTimeMillis())
                                    stats.fps = frameRate.ratePerSec(System.currentTimeMillis())
                                }
                            }
                            MsgType.PING -> {
                                val pong = Protocol.encodeMessage(MsgType.PONG, msg.body)
                                synchronized(sendLock) { output.write(pong); output.flush() }
                            }
                            MsgType.PONG -> if (stats != null && msg.body.size >= 8) {
                                stats.rttMs = (System.nanoTime() - bytesToLong(msg.body)) / 1_000_000.0
                            }
                            MsgType.AUDIO -> listener.onAudio(msg.body)
                            MsgType.OVERLAY -> listener.onOverlay(msg.body.isNotEmpty() && msg.body[0].toInt() != 0)
                            MsgType.BYE -> return
                            else -> { /* ignore */ }
                        }
                        msg = parser.next()
                    }
                } else if (n < 0) {
                    return  // peer closed
                }
            }
        } finally {
            out = null
        }
    }
}

File: client/src/transport_client.cpp
#include "transport_client.h"
#include "client_socket_channel.h"
#include "tls_trust.h"
#include "pairing_code.h"
#include <chrono>

namespace droppix {

std::optional<PairingProbe> probe_pairing_code(const std::string& host, uint16_t port,
                                               int timeout_ms) {
  auto ch = ClientSocketChannel::connect(host, port, /*use_tls=*/true, timeout_ms);
  if (!ch || !ch->peer_certificate()) return std::nullopt;
  auto der = cert_der(ch->peer_certificate());
  std::string fp = cert_fingerprint(ch->peer_certificate());
  ch->close();
  if (der.empty() || fp.empty()) return std::nullopt;
  return PairingProbe{derive_pairing_code(der), fp};
}

namespace {
std::vector<unsigned char> u64_be(uint64_t x) {
  std::vector<unsigned char> b(8);
  for (int i = 0; i < 8; ++i) b[i] = static_cast<unsigned char>(x >> (56 - i * 8));
  return b;
}
int64_t now_ms() {
  return std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::steady_clock::now().time_since_epoch()).count();
}
int64_t now_ns() {
  return std::chrono::duration_cast<std::chrono::nanoseconds>(
             std::chrono::steady_clock::now().time_since_epoch()).count();
}
}  // namespace

void TransportClient::close() {
  std::lock_guard<std::mutex> lk(sendLock_);
  channel_ = nullptr;
}

void TransportClient::sendTouch(const std::vector<TouchContact>& contacts) {
  std::lock_guard<std::mutex> lk(sendLock_);
  if (!channel_) return;
  auto msg = encode_message(MsgType::Touch, encode_touch(contacts));
  channel_->send_all(msg.data(), msg.size());
}

void TransportClient::sendOrientation(uint8_t code) {
  std::lock_guard<std::mutex> lk(sendLock_);
  if (!channel_) return;
  auto msg = encode_message(MsgType::Orientation, encode_orientation(code));
  channel_->send_all(msg.data(), msg.size());
}

void TransportClient::sendScroll(int dx, int dy, uint16_t x, uint16_t y) {
  std::lock_guard<std::mutex> lk(sendLock_);
  if (!channel_) return;
  auto msg = encode_message(MsgType::Scroll,
      encode_scroll(static_cast<int16_t>(dx), static_cast<int16_t>(dy), x, y));
  channel_->send_all(msg.data(), msg.size());
}

void TransportClient::sendMouseButton(uint8_t button, uint8_t action, uint16_t x, uint16_t y) {
  std::lock_guard<std::mutex> lk(sendLock_);
  if (!channel_) return;
  auto msg = encode_message(MsgType::MouseButton, encode_mouse_button(button, action, x, y));
  channel_->send_all(msg.data(), msg.size());
}

void TransportClient::sendKey(uint16_t keycode, uint8_t action) {
  std::lock_guard<std::mutex> lk(sendLock_);
  if (!channel_) return;
  auto msg = encode_message(MsgType::Key, encode_key(keycode, action));
  channel_->send_all(msg.data(), msg.size());
}

void TransportClient::runOverChannel(ByteChannel& channel, uint32_t width, uint32_t height,
                                    uint32_t density, uint32_t fps, uint8_t audio_wanted,
                                    uint8_t orientation_code, uint32_t bitrate_kbps,
                                    StreamListener& listener,
                                    const std::function<bool()>& isRunning,
                                    const std::string& name, const std::string& id,
                                    int pingIntervalMs) {
  {
    std::lock_guard<std::mutex> lk(sendLock_);
    channel_ = &channel;
    auto hello = encode_message(MsgType::Hello,
        encode_hello(kProtocolVersion, width, height, density, name, id,
                     fps, audio_wanted, orientation_code, bitrate_kbps));
    if (!channel.send_all(hello.data(), hello.size())) { channel_ = nullptr; return; }
  }

  MessageParser parser;
  std::vector<unsigned char> chunk(65536);
  int64_t lastPing = 0;

  while (isRunning()) {
    int64_t nowMs = now_ms();
    if (nowMs - lastPing >= pingIntervalMs) {
      std::lock_guard<std::mutex> lk(sendLock_);
      if (!channel_) break;
      auto ping = encode_message(MsgType::Ping, u64_be(static_cast<uint64_t>(now_ns())));
      if (!channel.send_all(ping.data(), ping.size())) break;
      lastPing = nowMs;
    }
    if (!channel.wait_readable(200)) continue;   // periodic wakeup to recheck isRunning()
    ssize_t n = channel.recv(chunk.data(), chunk.size());
    if (n == 0) continue;      // no data ready (non-blocking edge) — keep looping
    if (n < 0) break;          // peer closed / error
    parser.feed(chunk.data(), static_cast<size_t>(n));
    ParsedMessage msg;
    while (parser.next(msg)) {
      switch (msg.type) {
        case MsgType::Config: {
          uint32_t w, h, fps; std::vector<unsigned char> ed;
          if (decode_config(msg.body, w, h, fps, ed)) listener.onConfig(w, h, fps, ed);
          break;
        }
        case MsgType::Video: {
          uint64_t pts; bool key; std::vector<unsigned char> nal;
          if (decode_video(msg.body, pts, key, nal)) listener.onVideo(pts, key, nal);
          break;
        }
        case MsgType::Audio:
          listener.onAudio(msg.body);
          break;
        case MsgType::Overlay: {
          uint8_t show;
          if (decode_overlay(msg.body, show)) listener.onOverlay(show != 0);
          break;
        }
        case MsgType::Ping: {
          std::lock_guard<std::mutex> lk(sendLock_);
          if (!channel_) break;
          auto pong = encode_message(MsgType::Pong, msg.body);
          channel.send_all(pong.data(), pong.size());
          break;
        }
        case MsgType::Pong:
        case MsgType::Bye:
          if (msg.type == MsgType::Bye) { std::lock_guard<std::mutex> lk(sendLock_); channel_ = nullptr; return; }
          break;
        default:
          break;  // ignore
      }
    }
  }
  std::lock_guard<std::mutex> lk(sendLock_);
  channel_ = nullptr;
}

}  // namespace droppix
